community/index/src/main/java/org/neo4j/index/gbptree/GenSafePointerPair.java

```diff
@@ -0,0 +1,485 @@
+ * </ul>
+ * There are variations of pointers written in UNSTABLE generation:
+ * <ul>
+ * <li>BROKEN: written during a concurrent page cache flush and wasn't flushed after that point before crash</li>
+ * <li>CRASH: pointers written as UNSTABLE before a crash happened (non-clean shutdown) are seen as
+ * CRASH during recovery</li>
+ * </ul>
+ * <p>
+ * Combinations of above mentioned states of the two pointers dictates which, if any, to read from or write to.
+ * From the perspective of callers there's only "read" and "write", the two pointers are hidden.
+ * <p>
+ * All methods are static and all interaction is made with primitives.
+ * <p>
+ * Flags in results from read/write method calls. Pointer is 6B so msb 2B can be used for flags,
+ * although the most common case (successful read) has its flag zeros so a successful read doesn't need
+ * any masking to extract pointer.
+ * <pre>
+ * [    ,    ][    ,    ][ ... 6B pointer data ... ]
+ *  ▲▲▲▲ ▲▲▲▲  ▲▲▲
+ *  ││││ ││││  ││└────────────────────────────────────── 0:{@link #WRITE_TO_A}/1:{@link #WRITE_TO_B}
+ *  ││││ │││└──└└─────────────────────────────────────── POINTER STATE B
+ *  ││││ └└└──────────────────────────────────────────── POINTER STATE A
+ *  ││└└──────────────────────────────────────────────── GENERATION COMPARISON:
+ *  ││                                                   {@link #GEN_B_BIG}, {@link #GEN_EQUAL}, {@link #GEN_A_BIG}
+ *  │└────────────────────────────────────────────────── 0:{@link #READ}/1:{@link #WRITE}
+ *  └─────────────────────────────────────────────────── 0:{@link #SUCCESS}/1:{@link #FAIL}
+ * </pre>
+ */
+class GenSafePointerPair
+{
+    public static final int SIZE = GenSafePointer.SIZE * 2;
+
+    // Pointer states
+    static final byte STABLE = 0;     // any previous generation made safe by a checkpoint
+    static final byte UNSTABLE = 1;   // current generation, generation under evolution until next checkpoint
+    static final byte CRASH = 2;      // pointer written as unstable and didn't make it to checkpoint before crashing
+    static final byte BROKEN = 3;     // mismatching checksum
+    static final byte EMPTY = 4;      // generation and pointer all zeros
+
+    // Flags and failure information
+    static final long SUCCESS      = 0x00000000_00000000L;
+    static final long FAIL         = 0x80000000_00000000L;
+    static final long READ         = 0x00000000_00000000L;
+    static final long WRITE        = 0x40000000_00000000L;
+    static final long GEN_EQUAL    = 0x00000000_00000000L;
+    static final long GEN_A_BIG    = 0x10000000_00000000L;
+    static final long GEN_B_BIG    = 0x20000000_00000000L;
+    static final long WRITE_TO_A   = 0x00000000_00000000L;
+    static final long WRITE_TO_B   = 0x00200000_00000000L;
+    static final int STATE_SHIFT_A = 57;
+    static final int STATE_SHIFT_B = 54;
+
+    // Aggregations and masks
+    static final long SUCCESS_WRITE_TO_B = SUCCESS | WRITE | WRITE_TO_B;
```

Comment:
Yeah, I'm pretty sure some level of the compiler or JIT will remove those anyway. It looks great when reading the code.